File: EssentialProtect/src/main/java/essential/protect/Commands.java
package essential.protect;

import arc.Events;
import arc.util.Log;
import essential.bundle.Bundle;
import essential.event.CustomEvents;
import ksp.command.ClientCommand;
import mindustry.Vars;
import mindustry.gen.Player;
import mindustry.net.Administration;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Commands implementation for the EssentialProtect plugin.
 */
public class Commands {
    private static final String PLAYER_NOT_FOUND = "player.not.found";
    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    /**
     * Login command for player authentication.
     * This is a simplified version that doesn't interact with the database.
     *
     * @param playerData The player data
     * @param args The command arguments
     */
    @ClientCommand(name = "login", parameter = "<id> <password>", description = "Log-in to account.")
    public void login(essential.database.data.PlayerData playerData, String[] args) {
        if (args.length < 2) {
            playerData.err("command.login.usage");
            return;
        }

        Bundle bundle = playerData.getBundle();
        Player player = (Player) playerData.getPlayer();

        if (args[0].equals(args[1])) {
            player.sendMessage(bundle.get("command.login.same.password"));
            return;
        }

        // Simplified implementation that just shows a message
        Log.info("Login attempt: " + args[0]);
        player.sendMessage(bundle.get("command.login.not.found"));
    }

    /**
     * Register command for player account creation.
     * This is a simplified version that doesn't interact with the database.
     *
     * @param playerData The player data
     * @param args The command arguments
     */
    @ClientCommand(name = "reg", parameter = "<id> <password> <password_repeat>", description = "Register account")
    public void register(essential.database.data.PlayerData playerData, String[] args) {
        if (args.length < 3) {
            playerData.err("command.reg.usage");
            return;
        }

        Bundle bundle = playerData.getBundle();
        Player player = (Player) playerData.getPlayer();

        if (Main.conf.getAccount().getAuthType() != ProtectConfig.AuthType.None) {
            if (!args[1].equals(args[2])) {
                player.sendMessage(bundle.get("command.reg.incorrect"));
                return;
            }

            // Simplified implementation that just shows a message
            Log.info("Registration attempt: " + args[0]);
            player.sendMessage(bundle.get("command.reg.success"));
        } else if (Main.conf.getAccount().getAuthType() == ProtectConfig.AuthType.Discord) {
            if (Vars.mods.getMod("essential-discord") != null) {
                Events.fire(new CustomEvents.PlayerDiscordRequested(player.uuid()));
            } else {
                player.sendMessage(bundle.get("command.reg.unavailable-server"));
            }
        } else {
            player.sendMessage(bundle.get("command.reg.unavailable"));
        }
    }

    /**
     * Report command for reporting players.
     *
     * @param playerData The player data
     * @param args The command arguments
     */
    @ClientCommand(name = "report", parameter = "<player> <reason...>", description = "Report a player")
    public void report(essential.database.data.PlayerData playerData, String[] args) {
        if (args.length < 2) {
            playerData.err("command.report.usage");
            return;
        }

        Player player = (Player) playerData.getPlayer();
        Administration.PlayerInfo target = Vars.netServer.admins.findByName(args[0]).first();

        if (target != null) {
            String reason = args[1];
            String date = LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME);

            // Format the report text
            String text = new Bundle().get("command.report.texts", 
                target.lastName, player.plainName(), reason, 
                target.lastName, String.join(",", target.names), 
                target.id, target.lastIP, String.join(",", target.ips));

            // Log the report (simplified)
            Log.info("REPORT: " + date + " " + text);

            Log.info(new Bundle().get("command.report.received", 
                player.plainName(), target.lastName, reason));

            playerData.send("command.report.done", target.lastName);

            // Fire report event
            Events.fire(new CustomEvents.PlayerReported(
                player.plainName(), target.lastName, reason));
        } else {
            playerData.err(PLAYER_NOT_FOUND);
        }
    }

    /**
     * Clean up resources when the plugin is unloaded.
     */
    public void dispose() {
        executor.shutdown();
    }
}

File: EssentialProtect/src/main/kotlin/essential/protect/Commands.kt
package essential.protect

import arc.Core
import arc.Events
import arc.struct.ObjectSet
import arc.util.Log
import essential.bundle.Bundle
import essential.core.Main.Companion.scope
import essential.database.data.PlayerData
import essential.database.data.createPlayerData
import essential.database.data.mapToPlayerDataList
import essential.database.data.update
import essential.database.table.PlayerTable
import essential.event.CustomEvents
import essential.event.CustomEvents.PlayerDiscordRequested
import essential.event.CustomEvents.PlayerReported
import essential.log.LogType
import essential.log.writeLog
import essential.protect.Main.Companion.conf
import essential.util.currentTime
import essential.util.findPlayerData
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import ksp.command.ClientCommand
import mindustry.Vars
import mindustry.net.Administration.PlayerInfo
import org.jetbrains.exposed.sql.and
import org.jetbrains.exposed.sql.or
import org.jetbrains.exposed.sql.transactions.experimental.newSuspendedTransaction

class Commands {
    companion object {
        private const val PLAYER_NOT_FOUND = "player.not.found"
    }
    @ClientCommand(name = "login", parameter = "<id> <password>", description = "Log-in to account.")
    fun login(playerData: PlayerData, arg: Array<out String>) {
        val bundle = playerData.bundle
        val player = playerData.player
        if (arg[0] == arg[1]) {
            player.sendMessage(bundle["command.login.same.password"])
            return
        }

        scope.launch {
            val result = newSuspendedTransaction {
                PlayerTable.select(PlayerTable.accountPW, PlayerTable.accountPW).where {
                    (PlayerTable.accountID eq arg[0]) and (PlayerTable.accountPW eq arg[1])
                }.mapToPlayerDataList()
            }

            Core.app.post {
                if (result.isNotEmpty()) {
                    val playerData = result.first()

                    if (playerData.accountID == arg[1]) {
                        player.sendMessage(bundle["command.login.default.password"])
                    } else if (playerData.isConnected) {
                        player.sendMessage(bundle["command.login.already"])
                    } else {
                        if (findPlayerData(playerData.uuid) == null) {
                            Events.fire(CustomEvents.PlayerDataLoad(playerData))
                        } else {
                            player.sendMessage(bundle["command.login.already"])
                        }
                    }
                } else {
                    playerData.err("command.login.not.found")
                }
            }
        }
    }

    @ClientCommand(name = "reg", parameter = "<id> <password> <password_repeat>", description = "Register account")
    fun register(playerData: PlayerData, arg: Array<out String>) {
        val bundle = playerData.bundle
        val player = playerData.player
        
        if (conf.account.getAuthType() != ProtectConfig.AuthType.None) {
            if (arg.size != 3) {
                player.sendMessage(bundle["command.reg.usage"])
            } else if (arg[1] != arg[2]) {
                player.sendMessage(bundle["command.reg.incorrect"])
            } else {
                val exists = runBlocking {
                    PlayerTable.select(PlayerTable.accountID, PlayerTable.name).where {
                        (PlayerTable.accountID eq arg[0]) or (PlayerTable.name eq player.plainName())
                    }.empty()
                }
                if (exists) {
                    player.sendMessage(bundle["command.reg.exists"])
                } else {
                    scope.launch {
                        val playerData = createPlayerData(player.name(), player.uuid(), arg[0], arg[1])
                        playerData.permission = "user"
                        playerData.update()
                    }
                    Log.info(bundle["log.data_created", player.plainName()])
                }
            }
        } else if (conf.account.getAuthType() != ProtectConfig.AuthType.Discord) {
            if (Vars.mods.getMod("essential-discord") != null) {
                Events.fire(PlayerDiscordRequested(player.uuid()))
            } else {
                player.sendMessage(bundle["command.reg.unavailable-server"])
            }
        } else {
            player.sendMessage(bundle["command.reg.unavailable"])
        }
    }

    @ClientCommand(name = "report", parameter = "<player> <reason...>", description = "Report a player")
    fun report(playerData: PlayerData, arg: Array<out String>) {
        val player = playerData.player
        val target: ObjectSet<PlayerInfo?> = Vars.netServer.admins.findByName(arg[0])
        target.first()?.let {
            val reason = arg[1]
            val infos: PlayerInfo = Vars.netServer.admins.findByName(it.plainLastName()).first()
            val date = currentTime()
            val text: String? = Bundle()["command.report.texts", it.plainLastName(), player.plainName(), reason, infos.lastName, infos.names, infos.id, infos.lastIP, infos.ips]
            writeLog(LogType.Report, date + text, it.plainLastName())
            Log.info(Bundle()["command.report.received", player.plainName(), it.plainLastName(), reason])
            playerData.send("command.report.done", it.plainLastName())
            Events.fire(PlayerReported(player.plainName(), it.plainLastName(), reason))
        } ?: run {
            playerData.err(PLAYER_NOT_FOUND)
        }
    }
}
